node-uniffi/src/lib.rs

```diff
@@ -0,0 +1,313 @@
+    pub async fn get_header_by_hash(&self, hash: String) -> Result<String> {
+        let node = self.node.read().await;
+        let node = node
+            .as_ref()
+            .ok_or(LuminaError::network("Node not initialized"))?;
+        let hash = Hash::from_str(&hash).map_err(|e| LuminaError::invalid_hash(e.to_string()))?;
+        let header = node.get_header_by_hash(&hash).await?;
+        Ok(header.to_string())
+    }
+
+    /// Get a synced header for the block with a given height.
+    pub async fn get_header_by_height(&self, height: u64) -> Result<String> {
+        let node = self.node.read().await;
+        let node = node
+            .as_ref()
+            .ok_or(LuminaError::network("Node not initialized"))?;
+        let header = node.get_header_by_height(height).await?;
+        Ok(header.to_string())
+    }
+
+    /// Gets headers from the given heights range.
+    ///
+    /// If start of the range is undefined (None), the first returned header will be of height 1.
+    /// If end of the range is undefined (None), the last returned header will be the last header in the
+    /// store.
+    ///
+    /// Returns array of serialized ExtendedHeader strings.
+    pub async fn get_headers(
+        &self,
+        start_height: Option<u64>,
+        end_height: Option<u64>,
+    ) -> Result<Vec<String>> {
+        let node = self.node.read().await;
+        let node = node
+            .as_ref()
+            .ok_or(LuminaError::network("Node not initialized"))?;
+
+        let headers = match (start_height, end_height) {
+            (None, None) => node.get_headers(..).await,
+            (Some(start), None) => node.get_headers(start..).await,
+            (None, Some(end)) => node.get_headers(..=end).await,
+            (Some(start), Some(end)) => node.get_headers(start..=end).await,
+        }?;
+
+        Ok(headers.into_iter().map(|h| h.to_string()).collect())
+    }
+
+    /// Gets data sampling metadata for a height.
+    ///
+    /// Returns serialized SamplingMetadata string if metadata exists for the height.
+    pub async fn get_sampling_metadata(&self, height: u64) -> Result<Option<String>> {
+        let node = self.node.read().await;
+        let node = node
+            .as_ref()
+            .ok_or(LuminaError::network("Node not initialized"))?;
+
+        let metadata = node.get_sampling_metadata(height).await?;
+        Ok(metadata.map(|m| serde_json::to_string(&m).unwrap()))
+    }
+
+    /// Returns the next event from the node's event channel.
+    pub async fn events_channel(&self) -> Result<Option<NodeEvent>> {
+        let mut events_subscriber = self.events_subscriber.lock().await;
+        match events_subscriber.as_mut() {
+            Some(subscriber) => match subscriber.try_recv() {
+                Ok(event) => Ok(Some(event.event.into())),
+                Err(e) => Err(LuminaError::network(e.to_string())),
+            },
+            None => Err(LuminaError::network("Node is not running")),
```

Comment:
since this function is async already and you used an async mutex, wouldn't it be better to `await subscriber.recv()` and return a `Result<NodeEvent>`?

Iirc the this can only fail if publisher is dropped, meaning that node was stopped, so we could return here `LuminaError::NodeNotRunning` if this happens.

Reply:
good catch! ty